src/openloco/windows/stationwnd.cpp

```diff
@@ -62,12 +667,237 @@ namespace openloco::ui::windows::station
+            { cargo::widgets, widx::tab_cargo, &cargo::events, &cargo::enabledWidgets },
+            { cargo_ratings::widgets, widx::tab_cargo_ratings, &cargo_ratings::events, &common::enabledWidgets }
+        };
+
+        static void prepare_draw(window* self)
+        {
+            // Reset tab widgets if needed.
+            auto tabWidgets = tabInformationByTabOffset[self->current_tab].widgets;
+            if (self->widgets != tabWidgets)
+            {
+                self->widgets = tabWidgets;
+                self->init_scroll_widgets();
             }
+
+            // Activate the current tab.
+            self->activated_widgets &= ~((1 << widx::tab_station) | (1 << widx::tab_cargo) | (1 << widx::tab_cargo_ratings));
+            widx widgetIndex = tabInformationByTabOffset[self->current_tab].widgetIndex;
+            self->activated_widgets |= (1ULL << widgetIndex);
+
+            // Put station and town name in place.
+            auto station = stationmgr::get(self->number);
+            commonFormatArgs[0] = station->name;
+            commonFormatArgs[1] = station->town;
+            commonFormatArgs[2] = (station->flags & 15) + 230;
+
+            // Resize common widgets.
+            self->widgets[common::widx::frame].right = self->width - 1;
+            self->widgets[common::widx::frame].bottom = self->height - 1;
+
+            self->widgets[common::widx::caption].right = self->width - 2;
+
+            self->widgets[common::widx::close_button].left = self->width - 15;
+            self->widgets[common::widx::close_button].right = self->width - 3;
+
+            self->widgets[common::widx::panel].right = self->width - 1;
+            self->widgets[common::widx::panel].bottom = self->height - 1;
+        }
+
+        // 0x0048E5DF
+        static void text_input(window* self, widget_index callingWidget, char* input)
+        {
+            if (callingWidget != common::widx::caption)
+                return;
+
+            if (strlen(input) == 0)
+                return;
+
+            addr<0x009C68E8, string_id>() = string_ids::error_cant_rename_station;
+
+            uint32_t* buffer = (uint32_t*)input;
+            game_commands::do_11(1, self->number, 1, buffer[0], buffer[1], buffer[2]);
+            game_commands::do_11(1, 0, 2, buffer[3], buffer[4], buffer[5]);
+            game_commands::do_11(1, 0, 0, buffer[6], buffer[7], buffer[8]);
+        }
+        // 0x0048E6F1
+        static void update(window* self)
+        {
+            std::printf("common::update :%d\n", self->current_tab);
+            self->frame_no++;
+            self->call_prepare_draw();
+            WindowManager::invalidate(WindowType::station, self->number);
+        }
+
+        static void renameStationPrompt(window* self, widget_index widgetIndex)
+        {
+            auto station = stationmgr::get(self->number);
+            commonFormatArgs[2] = station->name;
```

Comment:
This is missing a few more arguments

Reply:
`textinput::open_textinput` does not accept args as there's no conversion from `const void*` to `void*`

Reply:
Are the arguments even used? `station->name` is already passed, so perhaps `nullptr` suffices here, too.

Reply:
They are used as the station is named twice. Once in the message and again in the Text Input.
![image](https://user-images.githubusercontent.com/56926365/80917382-64703300-8d56-11ea-855a-1002e527853c.png)

Reply:
This is what I have so far:
```commonFormatArgs[4] = station->name;
commonFormatArgs[5] = station->town;

textinput::open_textinput(self, string_ids::title_station_name, string_ids::prompt_type_new_station_name, station->name, widgetIndex, &commonFormatArgs); ```
Which produces this:
![image](https://user-images.githubusercontent.com/56926365/80917582-cda47600-8d57-11ea-8282-9bca4d590cc2.png)
I'm at a loss as to how to get the station->town into the text input as well as the station->name

Reply:
I'd expect one of the station string ids to be a `{STRINGID}{STRINGID}` string, so that it uses the town and name from the stack. The disassembly reads:
```
.text:0048E5FB                 mov     cx, word ptr (stations+2Ch)[edi] ; town name
.text:0048E602                 mov     word_112C836, cx
.text:0048E609                 mov     word ptr dword_112C82E+2, cx
.text:0048E610                 movzx   ecx, word ptr stations[edi]
.text:0048E617                 mov     word ptr dword_112C82E, cx
```

So it looks like the town name is pushed in two places:
0x112C836 => 0x112C836 - 0x112C826 = 16 => 16/2 => offset 8
0x112C82E + 2 = 0x112C830 => 0x112C830 - 0x112C826 = 10 => 10/2 => offset 5

Then, as you surmised, the station name is pushed to offset 4.

Reply:
I've looked at this for a while and I think all you need to do is pass for the value args a pointer to the station->town.

Reply:
The string_Id is `"Type new name for {STRINGID} station:"`

Reply:
Yep you are right
`textinput::open_textinput(self, string_ids::title_station_name, string_ids::prompt_type_new_station_name, station->name, widgetIndex, &station->town);`
is all that is required